src/modules/mc_att_control/RateControl/RateControl.cpp

```diff
@@ -0,0 +1,155 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file RateControl.cpp
+ */
+
+#include <RateControl.hpp>
+#include <px4_defines.h>
+
+using namespace matrix;
+
+void RateControl::setGains(const Vector3f &P, const Vector3f &I, const Vector3f &D)
+{
+	_gain_p = P;
+	_gain_i = I;
+	_gain_d = D;
+}
+
+void RateControl::setDTermCutoff(const float loop_rate, const float cutoff, const bool force)
+{
+	// only do expensive filter update if the cutoff changed
+	if (force || fabsf(_lp_filters_d.get_cutoff_freq() - cutoff) > 0.01f) {
+		_lp_filters_d.set_cutoff_frequency(loop_rate, cutoff);
+		_lp_filters_d.reset(_rate_prev);
+	}
+}
+
+void RateControl::setTPA(const Vector3f &breakpoint, const Vector3f &rate)
+{
+	_tpa_breakpoint = breakpoint;
+	_tpa_rate = rate;
+}
+
+void RateControl::setSaturationStatus(const MultirotorMixer::saturation_status &status)
+{
+	_mixer_saturation_positive[0] = status.flags.roll_pos;
+	_mixer_saturation_positive[1] = status.flags.pitch_pos;
+	_mixer_saturation_positive[2] = status.flags.yaw_pos;
+	_mixer_saturation_negative[0] = status.flags.roll_neg;
+	_mixer_saturation_negative[1] = status.flags.pitch_neg;
+	_mixer_saturation_negative[2] = status.flags.yaw_neg;
+}
+
+Vector3f RateControl::update(const Vector3f rate, const Vector3f rate_sp, const float dt, const bool landed,
+			     const float thrust_sp)
+{
+	Vector3f gain_p_tpa = _gain_p.emult(tpa_attenuations(_tpa_breakpoint(0), _tpa_rate(0), thrust_sp));
```

Comment:
One little optimization we should add is only updating TPA on thrust setpoint changes. At 1 kHz that alone saves more than half a percent of total cpu usage.

Reply:
Actually I plan to get rid of TPA all together. I just didn't want to pack that into this refactoring pr.

TPA is the dirty solution to the more quadratic than linear thrust output function of your ESC, motor, propeller. But it's far less acurate and up to tuning of each axis which basically no one does.

I think https://docs.px4.io/master/en/advanced_config/parameter_reference.html#THR_MDL_FAC the mixer/output level compensation with lnear combination of linear and quadratic function is the far better approach and I'm only using that one. Here's also a good example: https://discuss.px4.io/t/thrust-curve-compensation-through-thr-mdl-fac/13207

What we need to make sure is that this way of compensating is well documented and then I don't see a reason to keep TPA.

Reply:
Nice, I was wondering why we had both.

Reply:
>Nice, I was wondering why we had both.

Someone added it, someone used it, the other method was not documented for a long time. Noone removed it. I'll do that.

Now it's actually nicely documented 🎉  https://docs.px4.io/v1.9.0/en/config_mc/pid_tuning_guide_multicopter.html#thrust_curve
We could add a complete example for an even lower barrier.